### DISCUSSION TITLE
Local plugin permissions

### DISCUSSION THREAD

**Discussion by DimitrieStraoanu** [strapi/strapi#7599] [Legacy Topics]
Hello,

I've created a local plugin with some custom functionality and i need to display or hide the plugin from the admin menu based on user roles (super admin, editor or author). Could not find any clear documentation about this subject.  Can anyone help me with this?

**Comment by DimitrieStraoanu** [comment]
i found that adding this line of code 

```
menu: {
      pluginsSectionLinks: [
        {
          destination: `/plugins/${pluginId}`,
          icon,
          label: {
            id: `${pluginId}.plugin.name`,
            defaultMessage: name,
          },
          name,
          permissions: [
            { action: 'plugins::votes.read', subject: null },
            
          ],
        },
      ],
    },
```

to the plugins/my-plugin/admin/src/index.js hides the plugin from the admin menu but i'm just guesing here.

**Reply by soupette** [reply]
@DimitrieStraoanu the documentation you found is the way to set permissions in a link that needs to be added in the menu. We do it the same way in the [Content Type Builder plugin](https://github.com/strapi/strapi/blob/2957ad5b0d1fcfe33463bed7e7ea889a7ac78a63/packages/strapi-plugin-content-type-builder/admin/src/index.js#L58-L71). 

However, we lack documentation on how to check permissions in your plugin front-end files, so you will have to check out the core code for an example.
Basically you have :
- **2 components:**
  - [CheckPagePermissions](https://github.com/strapi/strapi/tree/master/packages/strapi-helper-plugin/lib/src/components/CheckPagePermissions) that works similarly to an protected Route and redirects the user to the Homepage if he tries to access it without having the right permission. We usually use it in the entry point of a plugin to make sure the plugin is not accessible even by typing the url. Here's an [example](https://github.com/strapi/strapi/blob/2957ad5b0d1fcfe33463bed7e7ea889a7ac78a63/packages/strapi-plugin-content-type-builder/admin/src/containers/App/index.js#L21)
  - [CheckPermissions](https://github.com/strapi/strapi/tree/master/packages/strapi-helper-plugin/lib/src/components/CheckPermissions), that can be used to display a child component if the user has the associated permission, if not the child component will be hidden. Here's an [example](https://github.com/strapi/strapi/blob/2957ad5b0d1fcfe33463bed7e7ea889a7ac78a63/packages/strapi-admin/admin/src/containers/MarketplacePage/PluginCard/index.js#L114-L121)

- **2 hook:** 
  - [useUserPermissions](https://github.com/strapi/strapi/tree/master/packages/strapi-helper-plugin/lib/src/hooks/useUserPermissions): that is very useful in order to check multiple permissions before sending requests for instance. Here's an [example](https://github.com/strapi/strapi/blob/2957ad5b0d1fcfe33463bed7e7ea889a7ac78a63/packages/strapi-admin/admin/src/containers/Webhooks/ListView/index.js#L28-L54)
  - [useUser](https://github.com/strapi/strapi/blob/master/packages/strapi-helper-plugin/lib/src/hooks/useUser/index.js), this one does not manage any permission check but it is useful if you want to retrieve the current user's permissions.

I hope this information will help you in your further developments! 

Good luck

**Reply by DimitrieStraoanu** [reply]
Thx for the guidelines !

**Comment by derrickmehaffy** [comment]
Hello,

Let me ping @JabStrapi and @Aurelsicoko about this, as it was a topic I had originally brought up with them related to the RBAC Community Edition vs Enterprise Edition. For the moment I don't have a good solution, but lets see what those two think.

**Comment by DimitrieStraoanu** [comment]
Ok, so i understand how to check if the user has the permission to access my plugin using the components and hooks @soupette showed here but my question is how to add the permission for my plugin to the **user role** since my plugin does not appear inside the /admin/settings/roles/<roleID> -> plugins tab.
![roles](https://user-images.githubusercontent.com/48469159/91286323-f92fee00-e796-11ea-9d3b-ac3bf212f162.jpg)
